src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs

```diff
@@ -182,15 +187,19 @@ public async Task<bool> TryGetAsync(ArrayBuilder<DiagnosticData> list, Cancellat
                     using var _1 = ArrayBuilder<StateSet>.GetInstance(out var syntaxAnalyzers);
                     using var _2 = ArrayBuilder<StateSet>.GetInstance(out var semanticSpanBasedAnalyzers);
                     using var _3 = ArrayBuilder<StateSet>.GetInstance(out var semanticDocumentBasedAnalyzers);
+                    var seenCompilerAnalyzer = false;
                     foreach (var stateSet in _stateSets)
                     {
                         if (!ShouldIncludeAnalyzer(stateSet.Analyzer, _shouldIncludeDiagnostic, _owner))
                             continue;
 
-                        if (!await TryAddCachedDocumentDiagnosticsAsync(stateSet, AnalysisKind.Syntax, list, cancellationToken).ConfigureAwait(false))
+                        var (includeSyntax, includeSemantic) = ProcessDiagnosticKinds(stateSet.Analyzer, _diagnosticKinds, ref seenCompilerAnalyzer);
```

Comment:
i really don't ike teh ref parameter.  seems trivial to just hoist this check into this loop if it's important that we only see one compiler-analyzer.

Reply:
So, this ref parameter is mainly a small perf optimization of avoid having to repeatedly invoke `IsCompilerAnalyzer()` on every analyzer, which currently involves invoking `analyzer.GetType().ToString()` for the lack of better alternatives: https://github.com/dotnet/roslyn/blob/6b02be2614212008d33307e916e37cd53dfb3d52/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/DiagnosticAnalyzerExtensions.cs#L7-L31

Generally all our hosts pass the compiler analyzer as the very first analyzer in the analyzer array to execute, so this logic basically avoids the redundant `IsCompilerAnalyzer()` redundant check for all the remaining analyzers. I can simplify the code to remove this ref param, but I felt the perf gain is more important than slightly unclear code. Would you be fine if I add a descriptive enough comment for having this param?